### ISSUE TITLE
How to bundle, import, and package separate modules

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#3458]
TL;DR skip to the table at the bottom to see how I'm suggesting we move things around

### Done
 - [x] #2680 @jeffbcross _refactor(http): move http into its own module as sibling of angular2_
 - [x] #2948 @jeffbcross _fix(typings): typings template should make module name at end of file dynamic_
 - [x] #3540 @jeffbcross _feat(typings): bundle typings should be able to declare references_
 - [x] #3590 @jeffbcross _feat(package): add "typescript" field with typings information_
 - [x] #3555 _feat(bundles): include all bundles in npm distributions_
 - [x] cancelled #3416 _refactor(di): move di into its own top-level module_
 - [x] cancelled #3415 _refactor(facade): move facade into separate top-level module_
 - [x] cancelled #3588 _feat(dist): include "core" modules in angular2 npm dist_

### In-Progress

 - [ ] #3713 @jeffbcross _chore: re-arrange sources to better reflect npm package and import paths_
 - [ ] #3714 @jeffbcross _chore: change bundle rules to generate one sfx module and several minimal modules_

### Unscheduled
 - [ ] #3423 _export type definitions for test_lib_
 - [ ] #3651 _docs: API reference should also include global/ambient object names included in sfx bundles_
 - [ ] #3712 _chore(zone): add zone.js to bundles and npm distribution_

# Goals
 * Simple, predictable, consistent package installation and consumption experience for end-users
 * Have minimal, logical bundles for production use (router bundle, core bundle, http bundle, full bundle with deps, etc)
 * Pave path toward eventually being able to separate modules into separate repositories with independent development lifecycles
 * Maximum symmetry between Dart and JavaScript publishing (naming and structure mostly)
 * Logical import and require paths for end users, without having to think about hierarchy or relationships of modules (import \{Http\} from 'angular/http' or require('angular/router'))

# Use Cases
 * User wants to create a quick code sample on plunker
 * User wants to create a production app with no build process but reasonable byte size
 * User wants to create a production app with a proprietary build process that creates minified bundles
 * User wants to create a production app on HTTP/2 server, using SystemJS loader, with no bundles, but minified modules that are loaded when needed.

# Characteristics of Modules
 * Is likely to be used outside of an angular app, without a dependency on Angular?
 * Is it a runtime requirement of angular2?
 * Is useful as a runtime utility for third-party Angular libraries?
 * Is it primarily a tool, not useful in an app at runtime?

Module                             | Used Outside | Required | Useful to Plugins | Tooling    |
-----------------------------------|--------------|----------|-------------------|------------|
Angular2                           |              | &#10003; |                   |            |
Forms                              |              |          | &#10003;          |            |
Facade                             |              | &#10003; | &#10003;          |            |
Router    		                     |              |          |                   |            |
Change Detection                   |              | &#10003; |                   |            |
Zone.js             			         | &#10003;     | &#10003; | &#10003;   			 |            |
Ts2Dart                            |              |          | &#10003;          | &#10003;   |
Dependency Injection        		   | &#10003;     | &#10003; | &#10003;          |            |
Http                               | &#10003;     |          | &#10003;          |            |
WebSocket, Storage, etc            | &#10003;     |          | &#10003;          |            |
Tactical                           | &#10003;     |          |                   |            |
Benchpress                         | &#10003;     |          |                   | &#10003;   |
Material                           |              |          |                   |            |
Test lib                           |              |          | &#10003;          |            |
NgUpgrade                          |              |          | &#10003;          |            |
Universal                          |              |          | &#10003;          |            |
React Native Renderer              |              |          | &#10003;          |            |
Workers                            |              |          | &#10003;          |            |

# Proposal
 * If it is **not Required** and **not Tooling**, it becomes a subdirectory of modules/angular2/ and gets its own bundle.
 * If it is **not Required**, it should be moved to the top-level of the angular2/modules directory,
 , i.e. angular2/modules/http
 * All bundles will be pushed to code.angularjs.org
 * All runtime modules that are currently part of the core repo should be published inside of the angular2 package on NPM and Pub (zone.js should also be added to the npm dist).

| Module                  | NPM Package Name      | Pub Package       | Bundle         | Bundle Global | Primary Runtime JS Import Namespace* | Target Source Home (repo / path)                       |
|-------------------------|-----------------------|-------------------|----------------|---------------|--------------------------------------|--------------------------------------------------------|
| Angular2/Core           | angular2              | angular2          | core.js        | ngCore        | angular2/core                        | angular/angular/modules/angular2/core                  |
| Forms                   | angular2              | angular2          | forms.js       | ngForms       | angular2/forms                       | angular/angular/modules/angular2/forms                 |
| Facade                  | angular2              | angular2          |                |               | angular2/core                        | angular/angular/modules/angular2/core/facade           |
| Router                  | angular2              | angular2          | router.js      | ngRouter      | angular2/router                      | angular/angular/modules/angular2/router                |
| Change Detection        | angular2              | angular2          |                |               | angular2/core                        | angular/angular/modules/angular2/core/change_detection |
| Zone.js                 | zone.js               |                   | zone.js        | ngZone        |                                      | angular/zone.js                                        |
| Ts2Dart                 | ts2dart               |                   |                |               |                                      | angular/ts2dart                                        |
| Dependency Injection    | di                    |                   |                |               | angular2/core                        | angular/angular/modules/angular2/core/di               |
| Http                    | ngHttp                | angular2          | http.js        | ngHttp        | angular2/http                        | angular/angular/modules/angular2/http                  |
| WebSocket, Storage, etc | ng...                 | ...               | ....js         | ng...         | angular2/...                         | angular/angular/modules/...                            |
| Tactical                | tactical              |                   | tactical.js    | ngTactical    |                                      | angular/tactical                                       |
| Benchpress              | benchpress            | benchpress        |                |               |                                      | angular/angular/modules/benchpress                     |
| Material                | angular-material2     | angular2_material | material2.js   | ngMaterial    | material2                            | angular/angular/modules/material2                      |
| Test lib                | angular2              | angular2          | test_lib.js    | ngTestLib     | angular2/test_lib                    | angular/angular/modules/angular2/test_lib              |
| NgUpgrade               | angular2              | angular2_upgrade  | upgrade.js     | ngUpgrade     | angular2/upgrade                     | angular/angular/modules/angular2/upgrade               |
| Universal               | angular2_universal    |                   | **             |               | angular2_universal                   | angular/universal                                      |
| React Native Renderer   | angular2_react_native |                   |                |               | angular2_react_native                | angular/react-native-render                            |
| Web Workers             | angular2_workers      | angular2_workers  | web-workers.js | ngWebWorkers  | angular/web-workers                  | angular/angular/modules/angular2/web-workers           |

\* Primary import namespaces when used in the context of an angular 2 app, as opposed to being used independently of Angular.
For example: `import {FORM_DIRECTIVES} from 'angular2/forms';`.
\** No bundle because only used in nodejs environment.

## Bundles

These will be the bundles that are exported:

 * angular2.js (includes all other bundles)
 * core.js
 * forms.js
 * http.js
 * router.js
 * web-worker.js
 * zone.js

And these are the extensions with which each bundle will be published to code.angularjs.org and npm. Bundles will be in the `bundles` folder of the npm package and will be published to code.angularjs.org as `code.angularjs.org/<version-number>/<bundle-name>`.

 * .js
 * .d.ts (typings are in separate `typings` folder in npm distribution)
 * .js.map
 * .min.js (Using uglify for now, eventually using angular/ts-minify)
 * .dev.js (runs in dev mode)
 * .dev.js.map
 * -testing.js
 * -testing.js.map
 * -testing.d.ts

The angular2 bundle will also get an sfx version, which will include dependencies like Rx, reflect-metadata, traceur-runtime, and will export a global `ng` object.

 * angular2.sfx.js
 * angular2.sfx.min.js
 * angular2.sfx.js.map
 * angular2.sfx.min.js.map

The `-testing` bundles will include testing utilities and mocks for the respective module. The core-testing.js will include the bulk of the testing utilities, leaving other testing bundles to mostly just include mocks.

**Comment by mlaval** [comment]
What about https://github.com/angular/angular/tree/master/modules/angular2/src/test_lib please? Is it part of one of the module listed?

**Comment by jeffbcross** [comment]
I should include angular 1 -> 2 upgrade process as use case to consider.

**Comment by jeffbcross** [comment]
@mlaval good question. Test lib and mocks should be covered here as well.

**Comment by mhevery** [comment]
I like this. :+1: 

`ng-upgrade` and `test` should be included as separate bundles (ala `form`)

**Comment by Koslun** [comment]
If you don't have a reason not to, I would also add [angular/universal](https://github.com/angular/universal), [angular/react-native-renderer](https://github.com/angular/react-native-renderer) and the future (or available?) Nativescript renderer to that table. The two latter in particular being modules that you likely wouldn't always want or require.

Either way, sounds great for all the reasons already mentioned.

**Comment by jeffbcross** [comment]
Thanks @Koslun, worth mentioning other projects like those, and we should definitely have consistent distribution strategies.

**Comment by jeffbcross** [comment]
I updated the issue with ng-universal, react-native-renderer, test lib, and ng-upgrade. I also added a column with the runtime import namespace for clarity.

@mhevery what do you think of the test lib and ng-upgrade outputs?

Also, I'm thinking there's more complexity and confusion than benefit in publishing to the angular namespace, i.e. `npm install @angular/http`. It only really buys us two things: 1) A simple module name by not having to include a namespace in the module name, and 2) Some small assurance to those installing/consuming packages that it's the legit angular package. I'll try to get some more insight from other team members and users on this point.

**Comment by jeffbcross** [comment]
@vsavkin right now the proposal is to move forms out of the angular2.js bundle, but still part of the angular2 distribution. Do you want to advocate for keeping forms in the bundle? This would potentially allow us to move Rx out of the bundle as well, since I don't think anything else in the core bundle would depend on it.

**Comment by gdi2290** [comment]
Personally I believe `@angular/angular2` should hold the angular opinions (angular way) on how everything should be glued together while `@angular/core` only has the classes without `@angular/di` and modules like `@angular/router`, `@angular/http`, (`@angular/forms` probably not) being separate from Di if possible

there needs to be a discussion later for universal distribution but for now we can just assume `@angular/server` or `@angular/universal` will work although server bootstrap and ServerDomRenderer (pending) could be moved to core. 


With universal the repo needs to be able to push to other packaging systems outside of the javascript/dart ecosystem such as scala/ruby/python. This is only to ensure the developer experience is consistent but also for another reason that will be mentioned in the repo. I see universal becoming more of a list of modules that interplay within the angular ecosystem and ensure a baselist experience but not bound by javascript, hence universal.

**Comment by gdi2290** [comment]
For Use Cases (perhaps even related to Migration):
In my experience, there are four stages to an web app in a repo(s) lifecycle

 0. hackable (es5)
  * Dev from other lang
  * Doesn't know code

 1. toy/example (es5ish)
  * plunkr
  * tutorials

 2. small/medium  (es5+)
  * small app without any 3rd party services and probably no build step
  * medium small team with many 3rd party services and build step

 3. large/enterprise (es5+, largely depends on architecture and team)
  * large you have a couple teams in the project one framework and possibly more than one repo
  * enterprise this is where you have many teams scoped to one feature/page can have many frameworks

A web app, imho, grows/matures with time and so does the abstraction level/technical debt. 

The best feature of Angular 1 was completely dominating the first three levels (0,1,2) and having a great transition stories between stage one->two->three but not so much for stage four. Angular 1 also stubley ensured a certain structure that allowed you to fall into the pit of success when moving to stage three (small/medium, thx to services/di/ui-router/controller/scope abstractions). The problem was that the subtle structure that Angular 1 suggested falls apart during the next stage (four). 

With Angular 2 there's a great foundation now that more patterns/concepts are pushed into the framework the fourth stage story is great. There's no great transition stories between each stages at the moment

With Rails you can kinda cli your way to victory with stage one-two but also keep in mind the internet has now become its documentation (copy error paste in google and see answers in stackoverflow). Rails also gets questionable in stage four 

That's only been my experience and my experience may not reflect everything about webdev so I could be wrong. With that said, I do propose the Use Cases/Migration could be thought up in a similar fashion comparing other frameworks if not from my examples

**Comment by mhevery** [comment]
@jeffbcross >>> what do you think of the test lib and ng-upgrade outputs? >>YES and YES

**Comment by yjbanov** [comment]
A couple of things:

* `di` package already exists on pub, so we'll likely need to publish it under a different name
* `web-workers` should have its own bundle, and renamed to `angular2_workers` (pub doesn't allow hyphens in package names)

**Comment by jeffbcross** [comment]
Thanks @yjbanov I updated the table with changes to di and web workers. I also added a column to the table indicating the global object name for each bundle. (i.e. the object users will interact with when loading bundles via \<script\>)

**Comment by IgorMinar** [comment]
We also have several di, change detection and routing strategies. does it makes sense to pull those out? how much weight do they add?

**Comment by mhevery** [comment]
I am thinking no on DI, change-detection, its part of core.

**Comment by jeffbcross** [comment]
@vsavkin are there DI strategies that are not being used by core? I'll have to look at router strategies to better understand that. But I think @IgorMinar's point falls more in line with (what I think is) the ideal distribution and module loading strategy, where end users bundle (or load) as few modules as their application actually needs, and don't use our generated bundles. If there are optional strategies, it would be beneficial to move those to their own modules to fall in line with the "ideal" I've described, even if we do still include those in the angular2 bundle.

**Comment by jeffbcross** [comment]
EDIT: I added text to the npm distribution bullet of the proposal per some offline discussion: "This includes the bundle version of the module, as well as all supported module formats."

**Comment by vsavkin** [comment]
>> This would potentially allow us to move Rx out of the core angular2.js bundle as well, since I don't think anything else in the core bundle would depend on it.

Component events are implemented using observables, so Rx will have to stay part of core.

>> are there DI strategies that are not being used by core?

There is nothing in DI that is not used by core.

**Comment by vsavkin** [comment]
I am not sure if forms should be a separate bundle. We are going to change the View decorator to contain the default set of directives that will be available to the user by default. If forms is a separate bundle, that set cannot contain NgModel.

**Comment by jeffbcross** [comment]
EDIT: I added a backlog with some of the open issues relating to this issue.

**Comment by gdi2290** [comment]
Can we also separate renderers with some sort of naming convention? For example on the Server you could also bootstrap using workers then there's also DomRenderer(client) and DomRenderer(server) aka ServerDomRenderer. For example, react is decoupling their renderer completely outside of react for other renderers such as three.js, and canvas. 
```es6
var React = require('react');
var ReactDOM = require('react-dom');
ReactDOM.render(<MyComponent />, node);
```
`@angular/angular` can still require `@angular/dom_renderer` which would be the default case while allowing others to configure angular in their own way

**Comment by jeffbcross** [comment]
> For example on the Server you could also bootstrap using workers
@gdi2290 Are you referring to an app that may run in a worker in the browser, but may also be rendered from a server? And if so, are you just wanting to establish a naming convention now so two renders inside the same namespace don't accidentally share a name, or is this already a problem?

**Comment by johnpapa** [comment]
If I am reading this right, we would install form npm each package, and for each we would do something like this ...

```
import {angular2} from 'angular/angular2';
import {forms} from 'angular/forms';
import {facade} from 'angular/facade';
import {zone} from 'angular/zone';
import {di} from 'angular/di';
import {http} from 'angular/http';
import {router} from 'angular/router';
```

The current bits make it confusing to know where everything is located and how to properly import them as an end user. Which also affects the tooling, as an end user of that tooling.

**Comment by jeffbcross** [comment]
@johnpapa good feedback. You're correct on the import module names, but these particular packages would all be in a single angular2 npm package. And the names you're importing are invalid. You would import actual classes or functions from separate modules like so:

```typescript
import {bootstrap} from 'angular2/angular2';
import {Http, httpInjectables} from 'angular2/http';
```

Di, zones, facade, and probably forms would be part of angular2/angular2. Most people would never import zones or facade, and would typically just import `bind` from di.

**Comment by johnpapa** [comment]
Thanks, I typed that kinda quick. It would be ideal to list the exports in each of those modules, at least the key ones, for docs today.

I understand modularity is still key here. There are decision points on what to break into different modules, and then what to export form each module. There is also a decision on what to include in the angular2 npm package. I believe the modules should be as small as possible, and your key points for driving separation are solid.

In your table, I see `di` as a proposed npm package. I am concerned about discoverability and clarity of the names. Perhaps consider a naming convention reserved for angular driven packages to make this clear. `ng2-di` for example. The style is not as important as the result.

I think this is key here:
>  end users bundle (or load) as few modules as their application actually needs, and don't use our generated bundles

**Comment by gdi2290** [comment]
@jeffbcross yeah sorry I mentioned two things there. With Workers that can probably stay the same name unless it's highly dependent on the browser. 

The goal I'm looking for is being able to do this with Fullstack Angular 2

```es6
class Model {
  find(id){ 
    return this.buisnessLogicAsync(id);
  }

  @Server
  _privateMethodAsync() {
    // client would compile to an RPC `this.http.post('/example/_privateMethodAsync')`
    // since you annotate this method as server only
    return Observable.return({}); 
  }
  
  @Worker
  businessLogicAsync(id) {
    // long running sync task
    return Observable.return(fibonacci(id));
  }

  @Native
  animation() {
    // native code for animations etc
    return Observable.return({}); 
  }
  
}
```
Where you would compile the methods for each environment (to hide the network layer) when you precompile your application. (you should also be able to combine `@Worker` and `@Server`)

For Renderer I think since there was a separation of application layer from render layer there is no reason why that the packages shouldn't be as well. I'm suggesting a naming convention for renderers  now so more example `@angular/dom_renderer` but there's a question of where should the environment code live? Should it live in the package or separate per environment. An example of this is [`ServerDomRenderer`](https://github.com/angular/universal/blob/8516fce0a6399f92ec2a5c414082c6a2b6665968/modules/server/src/server_dom_renderer.ts) which is 60 lines of code and the node version of [`Http`](https://github.com/angular/universal/blob/8516fce0a6399f92ec2a5c414082c6a2b6665968/modules/server/src/http_server.ts) which really could be one line of code (only replace BrowserXhr with NodeXhr)

**Comment by jeffbcross** [comment]
@gdi2290 could you open a separate issue to discuss that?

**Comment by jeffwhelpley** [comment]
The preboot component within Angular Universal can also be its own NPM module. It can be used outside Angular and has no downstream dependencies.

**Comment by jeffbcross** [comment]
@mhevery, @IgorMinar and I have been discussing the best solution to this internally, and have landed on this being the new hierarchy of source files, bundles, and imports. The ultimate goal is a simple mental model for users of the framework, while considering constraints such as byte size and use of some modules outside of the context of Angular 2.

I'll update the table in the original comment, but a summary of changes to previous thinking:

Source directory structure:

```
- modules/
--- angular2/
----- core/
----- forms/
----- http/...
----- router/...
----- web-worker/...
--- benchmarks/...
--- benchmarks_external/...
--- benchpress/...
--- examples/...
--- material2/...
--- rtts_assert/... (deprecated)
--- transforms/...
```

 * Everything inside the modules/ folder _could_ stand as its own npm package, though not all modules are useful as a standalone npm package.
 * Get rid of `src/` and `test/`directories (the dart tree should add these during build process). Unit test files will live next to source files.
 * Consolidate "core" parts into single "angular2/core" import
 * Facade and DI will be part of core, not exported separately from core for the near future*.
 * Each subdirectory of `modules/angular2` will have its own bundle. i.e. core.js, http.js, router.js.
 * Modules that should work in Angular 1 should internally not have any angular2-di-specific imports. In other words, they should be designed to work with _some_ DI system, but shouldn't import anything from the `di` module like `bind` or `Injector`. There should be files in the root of each module to represent an angular-2-di-friendly export, and an angular-1-di-friendly export, i.e. `http.ts` and `ng-1-http.ts`.
 * Not publishing anything to `@angular` npm namespace for the time being
 * There will be one "quick" `angular2` bundle for plunkers and less-byte-sensitive apps. The angular2 bundle will have an `sfx` version (described in more detail below) which will publish all bundles and dependencies of angular2, such as Rx and reflect-metadata, SystemJS, es6-module-loader. 

These will be the bundles that are exported:

 * angular2.js (includes all other bundles)
 * core.js
 * forms.js
 * http.js
 * router.js
 * web-worker.js

And these are the extensions with which each bundle will be published to code.angularjs.org and npm. Bundles will be in the `bundles` folder of the npm package and will be published to code.angularjs.org as `code.angularjs.org/<version-number>/<bundle-name>`.

 * .js
 * .d.ts (typings are in separate `typings` folder in npm distribution)
 * .js.map
 * .min.js (Using uglify for now, eventually using angular/ts-minify)
 * .dev.js (runs in dev mode)
 * .dev.js.map
 * -testing.js
 * -testing.js.map
 * -testing.d.ts

The angular2 bundle will also get an sfx version, which will include dependencies like Rx, reflect-metadata, traceur-runtime, and will export a global `ng` object.

 * angular2.sfx.js
 * angular2.sfx.min.js
 * angular2.sfx.js.map
 * angualr2.sfx.min.js.map

The `-testing` bundles will include testing utilities and mocks for the respective module. The core-testing.js will include the bulk of the testing utilities, leaving other testing bundles to mostly just include mocks.

*Note: @vsavkin insists that DI should be published as a separate npm package. I think it's best to keep the folder structure as proposed here (angular2/core/di), but treat DI as a special case and give it its own package.json and include it in the angular2 npm release process.

@rkirov makes a good point that we should be encouraging bundles as a short-term, quick way to get started with Angular2, and in the long-term should work on providing better recommendations for production-ready build setups. He also points out that having more bundles may be more confusing, for little gain of saved bytes. With our current SystemJS-based build system, router is still 66KB minified, and Http is similar, while angular2 minified is about 504KB. So overall there is still significant byte savings for certain use cases by having separate bundles. Having individual published bundles also provides some transparency into file size of the different angular2 modules.

**Comment by jpsfs** [comment]
Hi,

Sorry for digging up this issue.
On a37 forms was being export as "angular2/forms". Now (a40) it's back to "angular2/angular2". I don't see this change referenced on the changelog. Is this a decision or a regression?

Best regards,
